lib/client/src/client.rs

```diff
@@ -788,6 +898,142 @@ impl MangoClient {
+        account: &mut MangoAccountValue,
+    ) -> anyhow::Result<PreparedInstructions> {
+        let mut serum3_closable_order_market_index = None;
+
+        for p in account.all_serum3_orders() {
+            let open_orders_acc = self
+                .account_fetcher
+                .fetch_raw_account(&p.open_orders)
+                .await?;
+            let open_orders_bytes = open_orders_acc.data();
+            let open_orders_data: &serum_dex::state::OpenOrders = bytemuck::from_bytes(
+                &open_orders_bytes[5..5 + std::mem::size_of::<serum_dex::state::OpenOrders>()],
+            );
+
+            let is_closable = open_orders_data.free_slot_bits == u128::MAX
+                && open_orders_data.native_coin_total == 0
+                && open_orders_data.native_pc_total == 0;
+
+            if is_closable {
+                serum3_closable_order_market_index = Some(p.market_index);
+                break;
+            }
+        }
+
+        let first_closable_slot =
+            serum3_closable_order_market_index.expect("couldn't find any serum3 slot available");
+
+        let ixs = PreparedInstructions::from_single(
+            self.serum3_close_open_orders_instruction(first_closable_slot),
+            self.context.compute_estimates.health_cu_per_serum,
+        );
+
+        let first_closable_market = account.serum3_orders(first_closable_slot)?;
+        let (tk1, tk2) = (
+            first_closable_market.base_token_index,
+            first_closable_market.quote_token_index,
+        );
+        account.token_position_mut(tk1)?.0.decrement_in_use();
+        account.token_position_mut(tk2)?.0.decrement_in_use();
+        account.deactivate_serum3_orders(first_closable_slot)?;
+
+        Ok(ixs)
+    }
+
+    async fn find_existing_or_try_to_replace_token_positions(
+        &self,
+        account: &mut MangoAccountValue,
+        token_indexes: Vec<TokenIndex>,
+    ) -> anyhow::Result<Option<PreparedInstructions>> {
+        let mut ixs = PreparedInstructions::new();
+
+        for token_index in token_indexes {
+            let result = self
+                .find_existing_or_try_to_replace_token_position(account, token_index)
+                .await?;
+            if let Some(ix) = result {
+                ixs.append(ix);
+            }
+        }
+        if ixs.is_empty() {
+            return Ok(None);
+        }
+
+        Ok(Some(ixs))
+    }
+
+    async fn find_existing_or_try_to_replace_token_position(
+        &self,
+        account: &mut MangoAccountValue,
+        token_index: TokenIndex,
+    ) -> anyhow::Result<Option<PreparedInstructions>> {
+        let token_position_missing = account
+            .ensure_token_position(token_index)
+            .is_anchor_error_with_code(MangoError::NoFreeTokenPositionIndex.error_code());
+
+        if !token_position_missing {
+            return Ok(None);
+        }
+
+        let ixs = self.deactivate_first_active_unused_token(account).await?;
```

Comment:
the behavior here is a bit inconsistent:
- if the token position could be created directly on `account`, it's being done
- if it can't be created, it may close a different position but wil _not_ alter `account` to add the new position